### PR TITLE
feat: add unique id index to lmdb

### PR DESCRIPTION
**Pull request by StriderDM**
Description
---
Adds unique_id index to lmdb

### CODE REVIEW

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 1]
```diff
@@ -445,6 +449,16 @@ impl LMDBDatabase {
             "utxo_commitment_index",
         )?;
 
+        if let Some(unique_id) = &output.features.unique_id.clone() {
```
Comment (sdbondi) [inline base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:452]:
```suggestion
        if let Some(unique_id) = output.features.unique_id.as_ref() {
```

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 2]
```diff
@@ -567,6 +581,10 @@ impl LMDBDatabase {
             "utxo_commitment_index",
         )?;
 
+        if let Some(unique_id) = input.features.unique_id.clone() {
```
Comment (sdbondi) [inline base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:584]:
```suggestion
        if let Some(unique_id) = input.features.unique_id.as_ref() {
```

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 3]
```diff
@@ -841,6 +859,9 @@ impl LMDBDatabase {
                     output.commitment.as_bytes(),
                     "utxo_commitment_index",
                 )?;
+                if let Some(unique_id) = output.features.unique_id.clone() {
```
Comment (sdbondi) [inline base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:862]:
```suggestion
                if let Some(unique_id) = output.features.unique_id.as_ref() {
```

**Review by sdbondi — COMMENTED** [review]
utAck - so the idea is that the hash preimage is `asset namespace/class || the token`?

**Comment by delta1** [comment]
> utAck - so the idea is that the hash preimage is `asset namespace/class || the token`?

Sounds right, I think according to [proposed RFC] `(parent_pub_key, unique_id)` should be unique 

[proposed RFC]: https://github.com/tari-project/tari/pull/3340/files#diff-62c91e11620b9f0105f0c6418dbce59c6924086076081011ef45d38d61c05617R82

**Review by delta1 — CHANGES_REQUESTED** [review]
as per the [proposed RFC](https://github.com/tari-project/tari/pull/3407#issuecomment-933309055) the key for the index should be `parent_public_key || unique_id`

**Comment by delta1** [comment]
I'm happy to pick this up as part of the validations I'm doing @StriderDM 👍

**Comment by delta1** [comment]
Included in #3416 thanks @StriderDM !

**Pull request closed without merging**